src/components/tl/mlx5/tl_mlx5_team.c

```diff
@@ -169,40 +146,169 @@ static inline ucc_status_t ucc_tl_mlx5_a2a_team_test(ucc_base_team_t *team)
+                    ucc_tl_mlx5_clean_mcast_comm(tl_team->mcast->mcast_comm);
+                }
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE;
+            }
+
+            tl_debug(team->context->lib, "initialized tl team: %p: MCAST component is %s ALLTOALL component is %s",
+                    team, (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_READY)?"ENABLED":"DISABLED",
+                    (tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_READY)?"ENABLED":"DISABLED");
+        }
+
+        ucc_free(tl_team->local_status_array);
+        ucc_free(tl_team->global_status_array);
+
+        return UCC_OK;
+    }
+
+    ucc_assert(!tl_team->global_sync_in_progress);
+    
+    if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_INIT &&
+            tl_team->a2a_state == TL_MLX5_TEAM_STATE_INIT ) {
+        // check if ctx is ready for a2a and mcast
+        tl_team->local_status_array  = ucc_calloc(1, 2*sizeof(ucc_status_t));
+        tl_team->global_status_array = ucc_calloc(1, 2*sizeof(ucc_status_t));
+
+        tl_team->local_status_array[0] = tl_team->a2a_status.local;
+        tl_team->local_status_array[1] = tl_team->local_mcast_ctx_ready?UCC_OK:UCC_ERR_NO_RESOURCE;
+
+        status = ucc_service_allreduce(
+            core_team, tl_team->local_status_array, tl_team->global_status_array,
+            UCC_DT_INT32, 2, UCC_OP_MIN, subset, &tl_team->scoll_req);
+        if (status < 0) {
+            tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                     "failed to collect global status");
+            return status;
+        }
+        tl_team->global_sync_in_progress = 1;
+
+        return UCC_INPROGRESS;
+    }
 
     a2a_status = ucc_tl_mlx5_a2a_team_test(team);
     if (a2a_status < 0) {
-        tl_error(team->context->lib, "ALLTOALL tl team: %p creation failed %d",
-                 team, a2a_status);
+        tl_warn(team->context->lib, "ALLTOALL tl team: %p creation failed %d",
+                team, a2a_status);
         tl_team->a2a_state = TL_MLX5_TEAM_STATE_ALLTOALL_NOT_AVAILABLE;
     }
 
     if (tl_team->mcast_state != TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE) {
         mcast_status = ucc_tl_mlx5_mcast_team_test(team);
         if (mcast_status < 0) {
-            tl_error(team->context->lib, "MCAST tl team: %p creation failed %d",
-                     team, mcast_status);
+            tl_warn(team->context->lib, "MCAST tl team: %p creation failed %d",
+                    team, mcast_status);
             tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE;
         }
     }
 
     if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE &&
         tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_NOT_AVAILABLE) {
-        tl_error(team->context->lib, "unable to initialize tl team: %p", team);
-        return UCC_ERR_NO_RESOURCE;
+        tl_warn(team->context->lib, "unable to initialize tl team: %p", team);
+        goto global_sync_post;
```

Comment:
why do we need this goto?

Reply:
In this line, we know that both a2a and mcast are not available for this specific process. Therefore, there is nothing else that can be done for it. That is why in this line, we jump to the final stage where we tell all other processes our state.

Reply:
But this goto only jumps over this block:
```
if (UCC_OK != a2a_status || UCC_OK != mcast_status) {
    return UCC_INPROGRESS;
}
```
I'm confused: wouldn't we have the two status equal to UCC_OK? what happens if one component is available but not the other ?

Reply:
so before this new patch, we had the below code which was making sense, we needed to return UCC_ERR_NO_RESOURCE  immediately and avoid continuing and returning UCC_OK. 
With the new patch, we no longer need this here and this should be moved up where we know global status is not-avvailable for both components. In this case, mlx5 team create must return error.
```
    if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE &&
        tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_NOT_AVAILABLE) {
        tl_error(team->context->lib, "unable to initialize tl team: %p", team);
        return UCC_ERR_NO_RESOURCE;
    }